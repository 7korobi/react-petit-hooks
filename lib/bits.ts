const BITLIMIT = 31
const BITMASK = 2 ** 31 - 1

type Labels<T extends readonly string[]> = T[number][]
type BitsDic<T extends readonly string[], X> = {
  [key in T[number]]: X
}

export class BitsData<T extends readonly string[]> extends Array {
  0: number
  1: BitsShadow<T>
  constructor(value: number, shadow: BitsShadow<T>) {
    super(2)
    this[0] = value
    this[1] = shadow
  }
}

export class BitsShadow<T extends readonly string[]> {
  labels: T
  mask: number
  posi: BitsDic<T, number>
  nega: BitsDic<T, number>
  idx: BitsDic<T, number>

  constructor(labels: T) {
    if (BITLIMIT < labels.length) {
      throw new Error('too much bits.')
    }
    this.labels = labels
    this.mask = 0
    this.posi = {} as any
    this.nega = {} as any
    this.idx = {} as any
    labels.forEach((label: T[number]) => {
      this.posi[label] = this.idx[label] = 0
      this.nega[label] = BITMASK
    })
    labels.forEach((label: T[number], idx) => {
      const posi = 2 ** idx
      const nega = BITMASK & ~posi
      this.mask |= posi
      this.posi[label] |= posi
      this.nega[label] &= nega
      this.idx[label] || (this.idx[label] = idx)
    })
  }

  by(src: T | Labels<T>): number
  by(src: number): Labels<T>
  by(src: T | Labels<T> | number): Labels<T> | number {
    if ('number' === typeof src) {
      const labels: Labels<T> = []
      this.labels.forEach((label: T[number]) => {
        if (src & this.posi[label]) {
          labels.push(label)
        }
      })
      return labels
    } else {
      let n = 0
      src.forEach((label: T[number]) => {
        n |= (this.posi[label] || 0) as number
      })
      return n
    }
  }

  is(n: number): [BitsData<T>, BitsDic<T, number>] {
    const o: BitsData<T> = new BitsData<T>(n, this)
    const proxy = new Proxy(o, {
      get([value, shadow], label: T[number]) {
        return value & shadow.posi[label]
      },
      set(data, label: T[number], set: number | boolean | null) {
        const [value, shadow] = data
        let bits: number
        switch (set) {
          case null:
          case false:
            bits = 0
            break
          case true:
            bits = shadow.posi[label]
            break
          default:
            bits = shadow.posi[label] & (set << shadow.idx[label])
        }
        data[0] = value & shadow.nega[label] | bits
        return true
      },
      has([value, shadow], label: T[number]) {
        return !!shadow.idx[label]
      },
    }) as any
    return [o, proxy]
  }

  isOneBit(x: number) {
    return 0 === (x & (x - 1))
  }

  firstOut(x: number) {
    return x & (x - 1)
  }
  firstIn(x: number) {
    return x | (x + 1)
  }

  firstLinksOut(x: number) {
    return ((x | (x - 1)) + 1) & x
  }
  firstLinksIn(x: number) {
    return ((x & (x + 1)) - 1) | x
  }

  findBitOn(x: number) {
    return x & -x
  }
  findBitOff(x: number) {
    return ~x & (x + 1)
  }

  fillHeadsToOn(x: number) {
    return x | (x - 1)
  }

  fillHeadsToOff(x: number) {
    return x & (x + 1)
  }

  headsBitOff(x: number) {
    return ~x & (x - 1)
  }
  headsBitOn(x: number) {
    return ~(~x | (x + 1))
  }

  headsBitOffAndNextOn(x: number) {
    return x ^ (x - 1)
  }

  snoob(x: number) {
    const minbit = x & -x
    const ripple = x + minbit
    const ones = ((x ^ ripple) >>> 2) / minbit
    return ripple | ones
  }

  count(x: number) {
    let n
    n = (x >>> 1) & 0x77777777
    x = x - n
    n = (n >>> 1) & 0x77777777
    x = x - n
    n = (n >>> 1) & 0x77777777
    x = x - n
    x = (x + (x >>> 4)) & 0x0f0f0f0f
    x = x * 0x01010101
    return x >>> 24
  }

  reverse(x: number) {

  }
}

export function Bits<T extends readonly string[]>(labels: T) {
  return new BitsShadow(labels)
}
