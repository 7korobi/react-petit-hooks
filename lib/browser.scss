body {
  overscroll-behavior: none;
  pointer-events: none;
  user-select: none;

  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

#safe-area-measure {
  pointer-events: none;
  user-select: none;
  opacity: 0;
  visibility: hidden;

  z-index: -99999;

  width: 1%;
  height: 1%;
  margin-top: env(safe-area-inset-top);
  margin-right: env(safe-area-inset-right);
  margin-bottom: env(safe-area-inset-bottom);
  margin-left: env(safe-area-inset-left);
}