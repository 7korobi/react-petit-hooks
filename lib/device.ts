export const __SPEC__ = 'undefined' === typeof window
export const __BROWSER__ = !__SPEC__
export const __TEST__ = __SPEC__ || 'localhost' === location.hostname
