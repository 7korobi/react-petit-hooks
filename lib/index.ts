export * from './browser'
export * from './poll'
export * from './storage'
export * from './time'

