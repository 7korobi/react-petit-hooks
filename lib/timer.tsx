import { useState, useEffect, useRef } from 'react'
import { to_msec, to_relative_time_distance, to_tempo_bare, Tempo, to_tempo } from './time'

type TEMPO = [number, Tempo]
type TICK = [number, Tempo, string]
type RelativeTickOption<T> = {
  limit: string
  format(id: T): string
}
type RelativeTickData = {
  idAt: number
  limitAt: number
}

export function useTempo(size_str: string, zero_str: string = '0s') {
  const [[timerId, tempo], setTick] = useState<TEMPO>([] as any)
  useEffect(init, [])
  return [tempo]

  function init() {
    tick()
  }

  function tick() {
    const now = Date.now()
    const tempo = to_tempo(size_str, zero_str, now)
    const { timeout } = tempo
    const timerId = setTimeout(tick, timeout) as any
    setTick([timerId, tempo])
  }
}

export function useRelativeTick<T extends number | string | Date>(
  id: T,
  { limit, format }: RelativeTickOption<T>
): [string, Tempo] {
  const option = useRef<RelativeTickData>({} as any)
  const [[timerId, tempo, text], setTick] = useState([null as any, null as any, ''])
  useEffect(reset, [id, limit])
  return [text, tempo]

  function reset() {
    option.current.idAt = new Date(id).getTime()
    option.current.limitAt = to_msec(limit)
    if (timerId) {
      clearTimeout(timerId)
    }
    const newTimerId = tick()
    return () => {
      clearTimeout(newTimerId)
    }
  }

  function tick(): number {
    const { idAt, limitAt } = option.current
    const now = Date.now()
    const at = now - idAt
    const [, interval, , template] = to_relative_time_distance(at)
    const tempo = to_tempo_bare(interval, 0, at)
    const { now_idx, timeout } = tempo

    if (limitAt < at) {
      const timerId = null as any
      const text = format(id)
      setTick([timerId, tempo, text])
      return timerId
    }

    if (at < -limitAt) {
      const timerId = setTimeout(tick, timeout) as any
      const text = format(id)
      setTick([timerId, tempo, text])
      return timerId
    }

    const text = template.replace('%s', String(Math.abs(now_idx)))
    const timerId = setTimeout(tick, timeout) as any
    setTick([timerId, tempo, text])
    return timerId
  }
}
