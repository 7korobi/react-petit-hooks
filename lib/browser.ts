import { useEffect, useState, useReducer } from "react"
import { __BROWSER__ } from "./device"


declare global {
  interface Window {
    visualViewport: {
      offsetLeft: number
      offsetTop: number
      pageLeft: number
      pageTop: number
      width: number
      height: number
      scale: number
      addEventListener(type: "resize" | "scroll", cb: (e: Event) => void): void
      removeEventListener(
        type: "resize" | "scroll",
        cb: (e: Event) => void
      ): void
    }
  }
}

type SIZE = [number, number]

const vp = __BROWSER__
  ? window.visualViewport
  : { width: 0, height: 0, scale: 1 }

export const ViewBox: { size: SIZE; scale: number } = {
  size: [vp.width, vp.height],
  scale: vp.scale,
}

export function useInternet(): [boolean] {
  const [isOnline, setIsOnline] = useState(true)

  if (__BROWSER__) {
    useEffect(() => {
      window.addEventListener("offline", network_state)
      window.addEventListener("online", network_state)
      return () => {
        window.removeEventListener("offline", network_state)
        window.removeEventListener("online", network_state)
      }
    }, [])
  }
  return [isOnline]

  function network_state() {
    setIsOnline(window.navigator.onLine)
  }
}

export function useVisible(): [boolean] {
  const [isVisible, setIsVisible] = useState(true)

  if (__BROWSER__) {
    useEffect(() => {
      document.addEventListener("visibilitychange", visible_state)
      return () => {
        document.removeEventListener("visibilitychange", visible_state)
      }
    }, [])
  }
  return [isVisible]

  function visible_state() {
    setIsVisible("hidden" !== document.visibilityState)
  }
}

export function useKeyboard(): [] {
  if (__BROWSER__) {
    useEffect(() => {
      document.addEventListener("keyup", keyboard)
      return () => {
        document.removeEventListener("keyup", keyboard)
      }
    }, [])
  }

  return []
  function keyboard(e: KeyboardEvent) {
    console.log(e)
  }
}

export function useContextMenu(): [boolean, (isMenu: boolean) => void] {
  const [isMenu, setIsMenu] = useState(true)

  if (__BROWSER__) {
    useEffect(() => {
      const { style } = document.body
      if (isMenu) {
        style.setProperty("--menu-display", "block")
      } else {
        style.setProperty("--menu-display", "none")
      }
    }, [isMenu])

    useEffect(() => {
      document.addEventListener("contextmenu", deny)
      return () => {
        document.removeEventListener("contextmenu", deny)
      }
    }, [])
  }

  return [isMenu, setIsMenu]

  function deny(e: { preventDefault: () => void }) {
    e.preventDefault()
    setIsMenu(true)
  }
}

export function useViewport(): [number, number, number] {
  const [width, setWidth] = useState<number>(ViewBox.size[0])
  const [height, setHeight] = useState<number>(ViewBox.size[1])
  const [scale, setScale] = useState<number>(ViewBox.scale)

  if (__BROWSER__) {
    useEffect(() => {
      resize()
      window.visualViewport.addEventListener("resize", resize)
      window.addEventListener("orientationChange", resize)

      return () => {
        window.visualViewport.removeEventListener("resize", resize)
        window.addEventListener("orientationChange", resize)
      }
    }, [])
  }

  return [width, height, scale]

  function resize() {
    const { style } = document.body
    const { scale, width, height } = window.visualViewport
    ViewBox.scale = scale

    setScale(scale)
    if (scale === 1) {
      style.setProperty("--view-width", `${width}px`)
      style.setProperty("--view-height", `${height}px`)

      ViewBox.size = [width, height]
      setWidth(width)
      setHeight(height)
    }
  }
}

interface Axis {
  label(n: number | null, oldVal?: string): string | null
}

class AccelAxis implements Axis {
  constructor(
    public min: number,
    public max: number,
    public minus: string,
    public plus: string
  ) {}
  label(n: number | null, oldVal: string) {
    if (null === n) {
      return null
    }
    if (n < -4) {
      return this.minus
    }
    if (4 < n) {
      return this.plus
    }
    if (oldVal === this.minus && n < -1) {
      return this.minus
    }
    if (oldVal === this.plus && 1 < n) {
      return this.plus
    }
    return ""
  }
}
class RotateAxis implements Axis {
  constructor(
    public min: number,
    public max: number,
    public minus: string,
    public plus: string
  ) {}
  label(n: number | null, oldVal: string) {
    if (null === n) {
      return null
    }
    if (n < -140) {
      return this.minus
    }
    if (140 < n) {
      return this.plus
    }
    if (oldVal === this.minus && n < -30) {
      return this.minus
    }
    if (oldVal === this.plus && 30 < n) {
      return this.plus
    }
    return ""
  }
}
class CompassAxis implements Axis {
  constructor(public min: number, public max: number) {}
  label(n: number | null) {
    if (null === n) {
      return null
    }
    const n1 = Math.floor(n)
    const n2 = Math.floor((n * 60) % 60)
    const n3 = Math.floor((n * 3600) % 60)

    return `${n1}°${n2}′${n3}″`
  }
}
class GeoAxis implements Axis {
  constructor(
    public min: number,
    public max: number,
    public minus: string,
    public plus: string
  ) {}
  label(n: number | null) {
    if (null === n) {
      return null
    }
    const n1 = Math.floor(n)
    const n2 = Math.floor((n * 60) % 60)
    const n3 = Math.floor((n * 3600) % 60)
    const n4 = Math.floor((n * 216000) % 60)
    const n5 = Math.floor((n * 12960000) % 60)

    const mark = n < 0 ? this.minus : this.plus
    return `${n1}°${n2}′${n3}″${n4}‴${n5}⁗${mark}`
  }
}
class MksAxis implements Axis {
  constructor(public min: number, public max: number) {}
  label(n: number | null) {
    if (null === n) {
      return null
    }
    const n1 = Math.floor(n)
    const n2 = Math.floor((n * 100) % 100)
    return `${n1}ｍ${n2}㎝`
  }
}

// 運動座標系
const xAxis = new AccelAxis(-10, 10, "右", "左") // m/s2
const yAxis = new AccelAxis(-10, 10, "上", "下") // m/s2
const zAxis = new AccelAxis(-10, 10, "表", "裏") // m/s2

const arAxis = new RotateAxis(-360, 360, "俯下", "仰上") // degree/s
const brAxis = new RotateAxis(-360, 360, "左折", "右折") // degree/s
const grAxis = new RotateAxis(-360, 360, "左巻", "右巻") // degree/s

const altAxis = new MksAxis(0, 10) // m
const spdAxis = new MksAxis(0, 10) // m/s

// 地球座標系
const aAxis = new CompassAxis(0, 360) // degree
const bAxis = new GeoAxis(-180, 180, "左折", "右折") // degree
const gAxis = new GeoAxis(-90, 90, "左巻", "右巻") // degree

const latAxis = new GeoAxis(-180, 180, "E", "W")
const lonAxis = new GeoAxis(-90, 90, "N", "S")

const headAxis = new CompassAxis(0, 360) // degree

export function useDeviceOrientation<U>(
  rotate: (oldVal: U, newVal: number | null, axis: CompassAxis | RotateAxis) => U
): [[U, U, U], boolean] {
  const [alpha, setAlpha] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, aAxis)
  )
  const [beta, setBeta] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, bAxis)
  )
  const [gamma, setGamma] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, gAxis)
  )
  const [absolute, setAbsolute] = useState(true)

  if (__BROWSER__) {
    useEffect(() => {
      if (DeviceOrientationEvent.requestPermission) {
        DeviceOrientationEvent.requestPermission().then((ps) => {
          if ("granted" === ps) {
            window.addEventListener("deviceorientation", onOrientation)
          }
        })
      } else {
        window.addEventListener("deviceorientation", onOrientation)
      }
      return () => {
        window.addEventListener("deviceorientation", onOrientation)
      }
    }, [])
  }

  return [[alpha, beta, gamma], absolute]

  function onOrientation({
    alpha,
    beta,
    gamma,
    absolute,
  }: DeviceOrientationEvent) {
    setAlpha([alpha!, aAxis])
    setBeta([beta!, bAxis])
    setGamma([gamma!, gAxis])
    setAbsolute(absolute)
  }

  function chkRotate(
    oldVal: U,
    [newVal, axis]: [number | null, CompassAxis | RotateAxis]
  ): U {
    return rotate(oldVal, newVal, axis)
  }
}

export function useDeviceMotion<T, U>(
  motion: (oldVal: T, newVal: number | null, axis: AccelAxis) => T,
  rotate: (oldVal: U, newVal: number | null, axis: RotateAxis) => U
): [[T, T, T], [T, T, T], [T, T, T], [U, U, U], number] {
  const [ax, setAx] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, xAxis)
  )
  const [ay, setAy] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, yAxis)
  )
  const [az, setAz] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, zAxis)
  )

  const [gx, setGx] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, xAxis)
  )
  const [gy, setGy] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, yAxis)
  )
  const [gz, setGz] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, zAxis)
  )

  const [mx, setMx] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, xAxis)
  )
  const [my, setMy] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, yAxis)
  )
  const [mz, setMz] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, zAxis)
  )

  const [alpha, setAlpha] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, arAxis)
  )
  const [beta, setBeta] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, brAxis)
  )
  const [gamma, setGamma] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, grAxis)
  )
  const [interval, setInterval] = useState(0)

  if (__BROWSER__) {
    useEffect(() => {
      if (DeviceMotionEvent.requestPermission) {
        DeviceMotionEvent.requestPermission().then((ps) => {
          if ("granted" === ps) {
            window.addEventListener("devicemotion", onMotion)
          }
        })
      } else {
        window.addEventListener("devicemotion", onMotion)
      }
      return () => {
        window.addEventListener("devicemotion", onMotion)
      }
    }, [])
  }

  return [
    [ax, ay, az],
    [gx, gy, gz],
    [mx, my, mz],
    [alpha, beta, gamma],
    interval,
  ]

  function onMotion({
    interval,
    acceleration,
    accelerationIncludingGravity,
    rotationRate,
  }: DeviceMotionEvent) {
    const { x: ax, y: ay, z: az } = acceleration!
    const { x: mx, y: my, z: mz } = accelerationIncludingGravity!
    const { alpha, beta, gamma } = rotationRate!
    setAx([ax!, xAxis])
    setAy([ay!, yAxis])
    setAz([az!, zAxis])
    setMx([mx!, xAxis])
    setMy([my!, yAxis])
    setMz([mz!, zAxis])
    setGx([mx! - ax!, xAxis])
    setGy([my! - ay!, yAxis])
    setGz([mz! - az!, zAxis])
    setAlpha([alpha!, arAxis])
    setBeta([beta!, brAxis])
    setGamma([gamma!, grAxis])
    setInterval(interval)
  }

  function chkMotion(oldVal: T, [newVal, axis]: [number | null, AccelAxis]): T {
    return motion(oldVal, newVal, axis)
  }
  function chkRotate(
    oldVal: U,
    [newVal, axis]: [number | null, RotateAxis]
  ): U {
    return rotate(oldVal, newVal, axis)
  }
}

export function useGeoLocation<T, U>(
  geo: (oldVal: T, newVal: number | null, axis: GeoAxis | CompassAxis) => T,
  mks: (oldVal: U, newVal: number | null, axis: MksAxis) => U
): [[T, T], U, T, U] {
  const [latitude, setLatitude] = useReducer<typeof chkGeo>(
    chkGeo,
    geo(null!, null, latAxis)
  )
  const [longitude, setLongitude] = useReducer<typeof chkGeo>(
    chkGeo,
    geo(null!, null, lonAxis)
  )
  const [heading, setHeading] = useReducer<typeof chkCompass>(
    chkCompass,
    geo(null!, null, headAxis)
  )

  const [altitude, setAltitude] = useReducer<typeof chkMks>(
    chkMks,
    mks(null!, null, altAxis)
  )
  const [speed, setSpeed] = useReducer<typeof chkMks>(
    chkMks,
    mks(null!, null, spdAxis)
  )

  if (__BROWSER__) {
    if (!navigator?.geolocation) {
      return [[latitude, longitude], altitude, heading, speed]
    }

    useEffect(() => {
      const watch_id = navigator.geolocation.watchPosition(
        onPosition,
        ({ code }) => {
          console.log(`error watchPosition = ${code}`)
        },
        {
          enableHighAccuracy: true,
          maximumAge: 60 * 1000,
          timeout: 10 * 1000,
        }
      )
      return () => {
        navigator.geolocation.clearWatch(watch_id)
      }
    }, [])
  }

  return [[latitude, longitude], altitude, heading, speed]

  function onPosition({ coords, timestamp }: Position) {
    let { latitude, longitude, altitude, heading, speed } = coords
    setLongitude([longitude, lonAxis])
    setLatitude([latitude, latAxis])
    if (altitude !== null) setAltitude([altitude, altAxis])
    if (heading !== null) setHeading([heading, headAxis])
    if (speed !== null) setSpeed([speed, spdAxis])
  }
  function chkCompass(
    oldVal: T,
    [newVal, axis]: [number | null, CompassAxis]
  ): T {
    return geo(oldVal, newVal, axis)
  }
  function chkGeo(oldVal: T, [newVal, axis]: [number | null, GeoAxis]): T {
    return geo(oldVal, newVal, axis)
  }
  function chkMks(oldVal: U, [newVal, axis]: [number | null, MksAxis]): U {
    return mks(oldVal, newVal, axis)
  }
}
