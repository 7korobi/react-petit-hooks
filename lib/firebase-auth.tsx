import { useEffect, useState, createContext, useContext } from 'react'
import React from 'react'

import { User } from 'firebase/app'
import * as firebase from 'firebase/app'

if (__BROWSER__) {
  require('firebase/auth')
}

import { __BROWSER__ } from './device'

type ProviderProp = {
  children: React.ReactNode | React.ReactNode[]
}

type SetLoginProp = {
  signout?(): void
  facebook?(): void
  twitter?(): void
  github?(): void
  google?(): void
  microsoft?(): void
}

type FirebaseAuth = [
  User | null,
  firebase.auth.AdditionalUserInfo | null,
  firebase.auth.AuthCredential | null,
  SetLoginProp
]

export function useFirebaseAuth(): FirebaseAuth {
  const [user, setUser] = useState<firebase.User | null>(null)
  const [[additionalUserInfo, credential], setOption] = useState<
    [firebase.auth.AdditionalUserInfo | null, firebase.auth.AuthCredential | null]
  >([null, null])
  let setLogin
  if (user && credential) {
    setLogin = { signout }
  } else {
    setLogin = { facebook, twitter, github, google, microsoft }
  }
  const auth = firebase.auth()

  useEffect(() => {
    auth.onIdTokenChanged(setUser)
    auth.onAuthStateChanged(setUser)
    setPersistence().catch(onError)
    chkRedirected().catch(onError)
  }, [])

  return [user, additionalUserInfo, credential, setLogin]

  async function setPersistence(): Promise<void> {
    await auth.setPersistence(firebase.auth.Auth.Persistence.LOCAL)
    console.log('session persistence')
  }

  async function chkRedirected(): Promise<void> {
    const redirectResult = await auth.getRedirectResult()
    const { operationType, user, additionalUserInfo, credential } = redirectResult
    console.log('get redirect result success', redirectResult)

    setUser(user)
    setOption([additionalUserInfo || null, credential])
  }

  function onError({ code, message }: { code: string; message: string }) {
    console.warn(code, message)
  }

  function signout() {
    firebase.auth().signOut()
    setUser(null)
    setOption([null, null])
  }
  function facebook() {
    auth.signInWithRedirect(new firebase.auth.FacebookAuthProvider())
  }
  function twitter() {
    auth.signInWithRedirect(new firebase.auth.TwitterAuthProvider())
  }
  function github() {
    auth.signInWithRedirect(new firebase.auth.GithubAuthProvider())
  }
  function google() {
    auth.signInWithRedirect(new firebase.auth.GoogleAuthProvider())
  }
  function microsoft() {
    auth.signInWithRedirect(new firebase.auth.OAuthProvider('microsoft.com'))
  }
}

const FirebaseAuthContext = createContext<FirebaseAuth>([null, null, null, {}])

export function useAuth() {
  return useContext(FirebaseAuthContext)
}

export function FirebaseAuthProvider({ children }: ProviderProp) {
  const auth = useFirebaseAuth()
  return <FirebaseAuthContext.Provider value={auth}>{children}</FirebaseAuthContext.Provider>
}
