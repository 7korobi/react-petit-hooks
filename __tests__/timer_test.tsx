import { render, waitFor, act } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import React from 'react'
import format from 'date-fns/format'
import locale from 'date-fns/locale/ja'

import { useRelativeTick } from '../lib/timer'

test('after gap', () => {
  let test_now = new Date().getTime() + 60000
  const c = render(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 35000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 50000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 35000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 59 * 60000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 23 * 3600000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 6 * 24 * 3600000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 7 * 24 * 3600000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()
})

test('before gap', () => {
  let test_now = new Date().getTime() + 60020
  const c = render(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 35000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 50000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 35000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 59 * 60000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 23 * 3600000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 6 * 24 * 3600000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()

  test_now -= 7 * 24 * 3600000
  c.rerender(<TestData since={test_now} />)
  expect(c.container).toMatchSnapshot()
})

function TestData({ since }: { since: number }) {
  const [text] = useRelativeTick(since, {
    limit: '2d',
    format: (since) => format(since, 'yyyy/MM/dd(EE)é ƒ', { locale }),
  })
  return <>{text}</>
}
