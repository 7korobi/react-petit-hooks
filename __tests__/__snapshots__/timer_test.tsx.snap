// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`after gap 1`] = `
<div>
  60秒後
</div>
`;

exports[`after gap 2`] = `
<div>
  今
</div>
`;

exports[`after gap 3`] = `
<div>
  25秒前
</div>
`;

exports[`after gap 4`] = `
<div>
  1分前
</div>
`;

exports[`after gap 5`] = `
<div>
  1時間前
</div>
`;

exports[`after gap 6`] = `
<div>
  1日前
</div>
`;

exports[`after gap 7`] = `
<div>
  1週間前
</div>
`;

exports[`after gap 8`] = `
<div>
  2020/09/26(土)頃
</div>
`;

exports[`before gap 1`] = `
<div>
  2分後
</div>
`;

exports[`before gap 2`] = `
<div>
  26秒後
</div>
`;

exports[`before gap 3`] = `
<div>
  今
</div>
`;

exports[`before gap 4`] = `
<div>
  59秒前
</div>
`;

exports[`before gap 5`] = `
<div>
  59分前
</div>
`;

exports[`before gap 6`] = `
<div>
  23時間前
</div>
`;

exports[`before gap 7`] = `
<div>
  6日前
</div>
`;

exports[`before gap 8`] = `
<div>
  1週間前
</div>
`;
